src/codegen.rs

```diff
@@ -125,6 +134,86 @@ impl<'ink> CodeGen<'ink> {
+            PouIndexEntry::Function { name, linkage: LinkageType::Internal, is_generated: false, .. }
+            | PouIndexEntry::FunctionBlock { name, linkage: LinkageType::Internal, .. } => {
+                Some(String::from(name))
+            }
+            _ => None,
+        });
+        let all_names = global_index
+            .get_globals()
+            .values()
+            .map(VariableIndexEntry::get_qualified_name)
+            .map(String::from)
+            .chain(program_globals)
+            .chain(functions)
+            .map(|s| s.to_lowercase())
+            .map(|s| (crate::index::get_initializer_name(&s), s))
+            .fold(Vec::new(), |mut acc, (s, s1)| {
+                acc.push(s);
+                acc.push(s1);
+                acc
+            });
+
+        if let Some(got_entries) = &mut *got_layout.lock().unwrap() {
+            let mut new_symbols = Vec::new();
+            let mut new_got_entries = HashMap::new();
+            let mut new_got = HashMap::new();
+
+            for name in all_names {
+                if let Some(idx) = got_entries.get(&name.to_string()) {
+                    new_got_entries.insert(name.to_string(), *idx);
+                    index.associate_got_index(&name, *idx)?;
+                    new_got.insert(*idx, name.to_string());
+                } else {
+                    new_symbols.push(name.to_string());
+                }
+            }
+
+            // Put any names that weren't there last time in any free space in the GOT.
+            let mut idx: u64 = 0;
+            for name in &new_symbols {
+                while new_got.contains_key(&idx) {
+                    idx += 1;
+                }
+                new_got_entries.insert(name.to_string(), idx);
+                index.associate_got_index(name, idx)?;
+                new_got.insert(idx, name.to_string());
+            }
+
+            // Construct our GOT as a new global array. We initialise this array in the loader code.
+            let got_size: u32 = new_got
+                .keys()
+                .max()
+                .map_or(0, |m| *m + 1)
+                .try_into()
+                .expect("the computed custom GOT size is too large");
+
+            let ptr_ty = llvm.context.i8_type().ptr_type(AddressSpace::default());
+            let empty_got = ptr_ty
+                .const_array(vec![ptr_ty.const_null(); got_size as usize].as_slice())
+                .as_basic_value_enum();
+            let custom_got_ty =
+                BasicTypeEnum::ArrayType(Llvm::get_array_type(BasicTypeEnum::PointerType(ptr_ty), got_size));
+
+            let custom_got = llvm.create_global_variable(&self.module, "__custom_got", custom_got_ty);
+            custom_got.set_linkage(inkwell::module::Linkage::WeakODR);
+            custom_got.set_initial_value(Some(empty_got), custom_got_ty);
+
+            *got_entries = new_got_entries;
```

Comment:
As a reminder, the first and last lines here look a bit annoying because we are using a `Mutex`. This can be avoided by returning fragments of the custom GOT and merging them together at the end of the codegen pipeline before writing it to a file